datafusion/functions/src/unicode/substr.rs

```diff
@@ -89,29 +94,193 @@ impl ScalarUDFImpl for SubstrFunc {
+fn get_true_start_count(input: &str, start: usize, count: i64) -> (usize, usize) {
+    let (mut st, mut ed) = (input.len(), input.len());
+    let mut start_counting = false;
+    let mut cnt = 0;
+    for (char_cnt, (byte_cnt, _)) in input.char_indices().enumerate() {
+        if char_cnt == start {
+            st = byte_cnt;
+            if count != -1 {
+                start_counting = true;
+            } else {
+                break;
+            }
+        }
+        if start_counting {
+            if cnt == count {
+                ed = byte_cnt;
+                break;
+            }
+            cnt += 1;
+        }
+    }
+    (st, ed)
+}
+
+// The decoding process refs the trait at: arrow/arrow-data/src/byte_view.rs:44
+// From<u128> for ByteView
+fn string_view_substr(
+    string_view_array: &StringViewArray,
+    args: &[ArrayRef],
+) -> Result<ArrayRef> {
+    let mut builder = StringViewBuilder::new();
+    // Copy all blocks from input
+    for block in string_view_array.data_buffers() {
+        builder.append_block(block.clone());
+    }
+
+    let start_array = as_int64_array(&args[0])?;
+
+    match args.len() {
+        1 => {
+            for (idx, (raw, start)) in string_view_array
+                .views()
+                .iter()
+                .zip(start_array.iter())
+                .enumerate()
+            {
+                if let Some(start) = start {
+                    let length = *raw as u32;
+                    let start = (start - 1).max(0);
+
+                    // Operate according to the length of bytes
+                    if length == 0 {
+                        builder.append_null();
+                    } else if length > 12 {
+                        let view = ByteView::from(*raw);
+
+                        // Safety:
+                        // 1. idx < string_array.views.size()
+                        // 2. builder is guaranteed to have corresponding blocks
+                        unsafe {
+                            let str = string_view_array.value_unchecked(idx);
+                            let (start, end) =
+                                get_true_start_count(str, start as usize, -1);
+                            builder.append_view_unchecked(
+                                view.buffer_index,
+                                view.offset + start as u32,
+                                // guarantee that end-offset >= 0 for end <= str.len()
+                                (end - start) as u32,
+                            );
+                        }
+                    } else {
+                        // Safety:
+                        // (1) original bytes are valid utf-8,
+                        // (2) we do not slice on utf-8 codepoint
+                        unsafe {
+                            let bytes =
+                                StringViewArray::inline_value(raw, length as usize);
```

Comment:
I would expect that we never had to call `append_value` -- as I think the `substr` calculation can be done entirely in terms of the views (and never modify the values in the buffer). Reading this code carefully I do think is the case but it is quite implicit (it knows that calling append_value on a string that is less than 12 bytes) will not push values values into the buffer

I wonder if we could refactor some of this code into functions with names

Perhaps something like

```rust
/// Modify a `view` with length <= 12 bytes so it reflects the substring [start..end]
fn substr_small_view(len: usize, view: u128, start: usize, end: size) -> u128 { ... }

/// Modify a `view` with length > 12 bytes so it reflects the substring [start..end]
fn substr_large_view(len: usize, view: u128, start: usize, end: size) -> u128 { ... }
```

Then we could simply updated the views directly

However, the `StringViewBuilder` doesn't seem to allow for that at the moment....

Reply:
Maybe we could add a function like `append_view_u128_unchecked(view: u128)` in arrow/src/builder/generic_bytes_view_builder.rs to simply add a view with a given `u128`. Then the whole process would be:
1. Get the str of the view by `value_unchecked`, then get the [start, end)
2. `sub_view = if end-start>12 substr_large_view() else substr_small_view()`
3. call `appned_view_u128_unchecked(sub_view)` on the builder

Reply:
> Reading this code carefully I do think is the case but it is quite implicit (it knows that calling append_value on a string that is less than 12 bytes)

Yes

> Maybe we could add a function like append_view_u128_unchecked(view: u128)

I like this direction.

The current `append_view_unchecked` is basically only for long strings (as it asks for a block id), and we don't have a similar method for short strings, what's why we call `append_value` as a workaround.

We can probably have a function called `append_inlined_view_unchecked(len: usize, prefix: &[u8])` to make it the intention clearer

Reply:
The thing is that I think retrieving the `&str` by `value_unchecked()` is a must-do to calculate the `[start, end)`, then the `sub_view: u128` could be created based on the original view using `make_view`. I'll make some commits.